docs/source/design/progress_reporting.rst

```diff
@@ -0,0 +1,178 @@
+        'Unlocking files',
+        # optional unit string for a progress bar
+        unit=' Files',
+        # optional label to be displayed in a progress bar
+        label='Unlocking',
+        # maximum value for a progress bar
+        total=nfiles,
+    )
+
+A new progress bar will be created automatically for any report with a previously
+unseen activity ``identifier``. It can be configured via the specification of
+a number of arguments, most notably a target ``total`` for the progress bar.
+See :py:func:`datalad.log.log_progress` for a complete overview.
+
+Starting a progress report must be done with a dedicate call. It cannot be combined
+with a progress update.
+
+
+2. Update progress information about a process
+----------------------------------------------
+
+Any subsequent call to :py:func:`datalad.log.log_progress` with an activity
+identifier that has already been seen either updates, or finishes the progress
+reporting for an activity. Updates must contain an ``update`` key
+
+.. code-block:: python
+
+    log_progress(
+        lgr.info,
+        # must match the identier used to start the progress reporting
+        identifier,
+        # arbitrary message content, string expansion supported just like
+        # regular log messages
+        "Files to unlock %i", nfiles,
+        # critical key for report updates
+        update=1,
+        # ``update`` could be an absolute value or an increment
+        increment=True
+    )
+
+Updating a progress report can only be done after a progress reporting was
+initialized (see above).
+
+
+3. Report completion of a process
+---------------------------------
+
+A progress bar will remain active until it is explicitly taken down, even if an
+initially declared ``total`` value may have been reached. Finishing a progress
+report requires a final log message with the corresponding identifiers which
+does NOT contain an ``update`` key.
+
+.. code-block:: python
+
+    log_progress(
+        lgr.info,
+        identifier,
+        # closing log message
+        "Completed unlocking files",
+    )
+
+
+Progress reporting in non-interactive sessions
+----------------------------------------------
+
+:py:func:`datalad.log.log_progress` takes a `noninteractive_level` argument
+that can be sued to control how progress reporting happens when no
+progress bars can be used, but actual log messages are produced.
+
+.. code-block:: python
+
+    import logging
+
+    log_progress(
+        lgr.info,
+        identifier,
+        "Completed unlocking files",
+        noninteractive_level=logging.INFO
+    )
+
+Each call to :py:func:`~datalad.log.log_progress` can be given a different
+log level, in order to control the verbosity of the reporting in such a scenario.
+For example, it is possible to log the start or end of an activity at a higher
+level than intermediate updates. It is also possible to single out particular
+intermediate events, and report them at a higher level.
```

Comment:
and if no `noninteractive_level` is provided -- what happens? logs at some default level or doesn't log at all?

Reply:
It is always logged. Not sure if that is a good thing, but this is what it does right now.

Reply:
To be clear: always logged via whatever callable that was given to `log_progress()`. So whether or not it would show up, also depends on that choice. Our standard pattern is `lgr.info`

Reply:
Done.